modules/xfeatures2d/src/sift.cpp

```diff
@@ -627,7 +693,141 @@ static void calcSIFTDescriptor( const Mat& img, Point2f ptf, float ori, float sc
     cv::hal::magnitude32f(X, Y, Mag, len);
     cv::hal::exp32f(W, W, len);
 
-    for( k = 0; k < len; k++ )
+    k = 0;
+#if CV_AVX2
+    if( USE_AVX2 )
+    {
+        int CV_DECL_ALIGNED(32) idx_buf[8];
+        __m256 __ori = _mm256_set1_ps(ori);
+        __m256 __bins_per_rad = _mm256_set1_ps(bins_per_rad);
+        __m256i __n = _mm256_set1_epi32(n);
+        for( ; k <= len - 8; k+=8 )
+        {
+            __m256 __rbin = _mm256_loadu_ps(&RBin[k]);
+            __m256 __cbin = _mm256_loadu_ps(&CBin[k]);
+            __m256 __obin = _mm256_mul_ps(_mm256_sub_ps(_mm256_loadu_ps(&Ori[k]), __ori), __bins_per_rad);
+            __m256 __mag = _mm256_mul_ps(_mm256_loadu_ps(&Mag[k]), _mm256_loadu_ps(&W[k]));
+
+            __m256 __r0 = _mm256_floor_ps(__rbin);
+            __rbin = _mm256_sub_ps(__rbin, __r0);
+            __m256 __c0 = _mm256_floor_ps(__cbin);
+            __cbin = _mm256_sub_ps(__cbin, __c0);
+            __m256 __o0 = _mm256_floor_ps(__obin);
+            __obin = _mm256_sub_ps(__obin, __o0);
+
+            __m256i __o0i = _mm256_cvtps_epi32(__o0);
+            // _o0 += (o0 < 0) * n
+            __o0i = _mm256_add_epi32(__o0i, _mm256_and_si256(__n, _mm256_cmpgt_epi32(_mm256_setzero_si256(), __o0i)));
+            __o0i = _mm256_sub_epi32(__o0i,
+                _mm256_and_si256(__n, _mm256_or_si256(_mm256_cmpeq_epi32(__o0i, __n), _mm256_cmpgt_epi32(__o0i, __n))));
+
+            __m256 __v_r1 = _mm256_mul_ps(__mag, __rbin);
+            __m256 __v_r0 = _mm256_sub_ps(__mag, __v_r1);
+
+            __m256 __v_rc11 = _mm256_mul_ps(__v_r1, __cbin);
+            __m256 __v_rc10 = _mm256_sub_ps(__v_r1, __v_rc11);
+
+            __m256 __v_rc01 = _mm256_mul_ps(__v_r0, __cbin);
+            __m256 __v_rc00 = _mm256_sub_ps(__v_r0, __v_rc01);
+
+            __m256 __v_rco111 = _mm256_mul_ps(__v_rc11, __obin);
+            __m256 __v_rco110 = _mm256_sub_ps(__v_rc11, __v_rco111);
+
+            __m256 __v_rco101 = _mm256_mul_ps(__v_rc10, __obin);
+            __m256 __v_rco100 = _mm256_sub_ps(__v_rc10, __v_rco101);
+
+            __m256 __v_rco011 = _mm256_mul_ps(__v_rc01, __obin);
+            __m256 __v_rco010 = _mm256_sub_ps(__v_rc01, __v_rco011);
+
+            __m256 __v_rco001 = _mm256_mul_ps(__v_rc00, __obin);
+            __m256 __v_rco000 = _mm256_sub_ps(__v_rc00, __v_rco001);
+
+            __m256i __one = _mm256_set1_epi32(1);
+            __m256i __idx = _mm256_add_epi32(
+                _mm256_mullo_epi32(
+                    _mm256_add_epi32(
+                        _mm256_mullo_epi32(_mm256_add_epi32(_mm256_cvtps_epi32(__r0), __one), _mm256_set1_epi32(d + 2)),
+                        _mm256_add_epi32(_mm256_cvtps_epi32(__c0), __one)),
+                    _mm256_set1_epi32(n + 2)),
+                __o0i);
+
+            _mm256_store_si256((__m256i *)idx_buf, __idx);
+
+            hist[idx_buf[0]] +=               __v_rco000[0];
```

Comment:
MSVC compiler stops in this line (same problem with lines below too):
```
error C2676: binary '[': '__m256' does not define this operator or a conversion to a type acceptable to the predefined operator [C:\build\precommit-contrib_windows64\build\modules\xfeatures2d\opencv_xfeatures2d.vcxproj]
```
http://pullrequest.opencv.org/buildbot/builders/precommit-contrib_windows64/builds/10121